### PR TITLE
Added orders logic 

### PR DESCRIPTION
**Pull request by f3ops3**
Added orders logic by implementing services, repositories, models, mappers and controller classes. Added "/api" route

### CODE REVIEW

File: src/main/java/springweb/courseproject/controller/OrderController.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    public OrderResponseDto updateOrder(
+            @PathVariable Long orderId,
+            @RequestBody @Valid PatchOrderRequestDto patchOrderRequestDto) {
+        return orderService.patchOrder(orderId, patchOrderRequestDto);
+    }
+
+    @PreAuthorize("hasRole('USER')")
+    @Operation(summary = "Create a new order",
+            description = "Create a new order with current shopping cart")
+    @PostMapping
```
Comment (JJJazl) [inline src/main/java/springweb/courseproject/controller/OrderController.java:44]: Please, add corresponding status code for POST method

File: src/main/java/springweb/courseproject/controller/OrderController.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+@RestController
+@RequestMapping("/orders")
+@RequiredArgsConstructor
+public class OrderController {
+    private final OrderService orderService;
+
+    @PreAuthorize("hasRole('ADMIN')")
+    @Operation(summary = "Update an order", description = "Update status of specific order")
+    @PatchMapping("/{orderId}")
+    public OrderResponseDto updateOrder(
```
Comment (JJJazl) [inline src/main/java/springweb/courseproject/controller/OrderController.java:35]:
```suggestion
    public OrderResponseDto updateOrder(

```

File: src/main/java/springweb/courseproject/dto/order/PatchOrderRequestDto.java [hunk 3]
```diff
@@ -0,0 +1,11 @@
+package springweb.courseproject.dto.order;
+
+import jakarta.validation.constraints.NotNull;
+import lombok.Data;
+import springweb.courseproject.model.Status;
+
+@Data
+public class PatchOrderRequestDto {
```
Comment (JJJazl) [inline src/main/java/springweb/courseproject/dto/order/PatchOrderRequestDto.java:8]:
```suggestion
public class UpdateOrderRequestDto {

```

File: src/main/java/springweb/courseproject/service/order/OrderService.java [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+import springweb.courseproject.dto.order.OrderResponseDto;
+import springweb.courseproject.dto.order.PatchOrderRequestDto;
+import springweb.courseproject.dto.orderitem.OrderItemResponseDto;
+
+public interface OrderService {
+    OrderResponseDto createOrder(Long userId, CreateOrderRequestDto createOrderRequestDto);
+
+    List<OrderResponseDto> getAllUsersOrders(Pageable pageable, Long userId);
+
+    OrderResponseDto patchOrder(Long orderId, PatchOrderRequestDto patchOrderRequestDto);
```
Comment (JJJazl) [inline src/main/java/springweb/courseproject/service/order/OrderService.java:15]:
```suggestion
    OrderResponseDto updateOrder(Long orderId, PatchOrderRequestDto patchOrderRequestDto);

```

File: src/main/java/springweb/courseproject/service/order/OrderServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+        order.setOrderDate(LocalDateTime.now());
+        order.setShippingAddress(createOrderRequestDto.getShippingAddress());
+        order.setOrderItems(cart.getCartItems().stream()
+                .map(cartItem -> {
+                    OrderItem orderItem = orderItemMapper.toOrderItem(cartItem);
+                    orderItem.setOrder(order);
+                    return orderItem;
+                })
+                .collect(Collectors.toSet())
+        );
```
Comment (JJJazl) [inline src/main/java/springweb/courseproject/service/order/OrderServiceImpl.java:59]:
```suggestion
Set<OrderItems> orderItems = cart.getCartItems().stream()
                .map(cartItem -> {
                    OrderItem orderItem = orderItemMapper.toOrderItem(cartItem);
                    orderItem.setOrder(order);
                    return orderItem;
                })
                .collect(Collectors.toSet())
        order.setOrderItems(orderItems);
```

File: src/main/java/springweb/courseproject/service/order/OrderServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+        order.setTotal(cart.getCartItems().stream()
+                .map(cartItem -> cartItem.getBook().getPrice().multiply(
+                        BigDecimal.valueOf(cartItem.getQuantity())))
+                .reduce(BigDecimal.ZERO, BigDecimal::add)
+        );
```
Comment (JJJazl) [inline src/main/java/springweb/courseproject/service/order/OrderServiceImpl.java:64]:
```suggestion
BigDecimal total = cart.getCartItems().stream()
                .map(cartItem -> cartItem.getBook().getPrice().multiply(
                        BigDecimal.valueOf(cartItem.getQuantity())))
                .reduce(BigDecimal.ZERO, BigDecimal::add);
        order.setTotal(total);
```

**Review by JJJazl — CHANGES_REQUESTED** [review]

File: src/main/java/springweb/courseproject/service/order/OrderServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+public class OrderServiceImpl implements OrderService {
+    private final OrderRepository orderRepository;
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final OrderItemMapper orderItemMapper;
+    private final OrderMapper orderMapper;
+
+    @Override
+    public OrderResponseDto createOrder(Long userId, CreateOrderRequestDto createOrderRequestDto) {
+        ShoppingCart cart = shoppingCartRepository.findByUserId(userId).orElseThrow(
+                () -> new EntityNotFoundException("Shopping cart with user id "
```
Comment (liliia-ponomarenko) [inline src/main/java/springweb/courseproject/service/order/OrderServiceImpl.java:38]: this exception should return a 404 response status

File: src/main/java/springweb/courseproject/service/order/OrderServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+                        + userId + " not found")
+        );
+        if (cart.getCartItems().isEmpty()) {
+            throw new DataProcessingException("You can't place an order with empty shopping cart");
```
Comment (liliia-ponomarenko) [inline src/main/java/springweb/courseproject/service/order/OrderServiceImpl.java:42]: the same 400 response status should be return

File: src/main/java/springweb/courseproject/service/order/OrderServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+@Service
+@RequiredArgsConstructor
+public class OrderServiceImpl implements OrderService {
+    private final OrderRepository orderRepository;
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final OrderItemMapper orderItemMapper;
+    private final OrderMapper orderMapper;
+
+    @Override
+    public OrderResponseDto createOrder(Long userId, CreateOrderRequestDto createOrderRequestDto) {
```
Comment (liliia-ponomarenko) [inline src/main/java/springweb/courseproject/service/order/OrderServiceImpl.java:36]: Let’s  use Transactional

File: src/main/java/springweb/courseproject/service/order/OrderServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+                    orderItem.setOrder(order);
+                    return orderItem;
+                })
+                .collect(Collectors.toSet());
+        order.setOrderItems(orderItemSet);
+        BigDecimal total = cart.getCartItems().stream()
+                .map(cartItem -> cartItem.getBook().getPrice().multiply(
+                        BigDecimal.valueOf(cartItem.getQuantity())))
+                .reduce(BigDecimal.ZERO, BigDecimal::add);
+        order.setTotal(total);
```
Comment (liliia-ponomarenko) [inline src/main/java/springweb/courseproject/service/order/OrderServiceImpl.java:61]: Let’s  move it to the private method

File: src/main/java/springweb/courseproject/service/order/OrderServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+        return orderMapper.toDto(orderRepository.save(order));
```
Comment (liliia-ponomarenko) [inline src/main/java/springweb/courseproject/service/order/OrderServiceImpl.java:62]: Don’t  forget to clear the user's shopping cart

File: src/main/java/springweb/courseproject/service/order/OrderServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+        order.setStatus(updateOrderRequestDto.getStatus());
+        return orderMapper.toDto(orderRepository.save(order));
+    }
+
+    @Override
+    public List<OrderItemResponseDto> getOrderItemsById(Long orderId, Long userId) {
+        Order order = orderRepository.findById(orderId).orElseThrow(
+                () -> new EntityNotFoundException("Order with order id " + orderId + " not found")
+        );
+        checkIfOrderBelongsToUser(order, userId);
```
Comment (liliia-ponomarenko) [inline src/main/java/springweb/courseproject/service/order/OrderServiceImpl.java:86]: You can check on request level 
```suggestion
        Order order = orderRepository.findByIdAndUserId(orderId, userId).orElseThrow(
                () -> new EntityNotFoundException("Order with order id " + orderId + " not found...")
        );
```

File: src/main/java/springweb/courseproject/service/order/OrderServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+                .collect(Collectors.toList());
+    }
+
+    @Override
+    public OrderItemResponseDto getOrderItemFromOrderByIds(Long orderId,
+                                                           Long userId, Long orderItemId) {
+        Order order = orderRepository.findById(orderId).orElseThrow(
+                () -> new EntityNotFoundException("Order with order id " + orderId + " not found")
+        );
+        checkIfOrderBelongsToUser(order, userId);
```
Comment (liliia-ponomarenko) [inline src/main/java/springweb/courseproject/service/order/OrderServiceImpl.java:98]: the same

File: src/main/java/springweb/courseproject/service/order/OrderServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+        Optional<OrderItem> first = order.getOrderItems().stream()
+                .filter(item -> item.getId().equals(orderItemId))
+                .findFirst();
```
Comment (liliia-ponomarenko) [inline src/main/java/springweb/courseproject/service/order/OrderServiceImpl.java:101]: Can you refactor your query and avoid using this filter? (Find order item by id)

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s improve your solution ;)

**Review by liliia-ponomarenko — COMMENTED** [review]
One more thing, I noticed you link to the mate academy chekcstyle plugin. Add the checks file to your project directly.

**Review by okuzan — APPROVED** [review]

**Pull request merged**